### PR TITLE
GH-36420: [C++] Add An Enum Option For SetLookup Options

### LINKED ISSUE #36420: [C++] Add an sql-compatible is_in variant
### Describe the enhancement requested

This could be a new compute function or it could be new options for the existing is_in function.  Currently, the way `is_in` treats `null` does not match SQL semantics which makes it difficult to use when interoperability is expected.

The current behavior is summed up nicely in the `SetLookupOptions::skip_nulls` description:

```
/// Whether nulls in `value_set` count for lookup.
///
/// If true, any null in `value_set` is ignored and nulls in the input
/// produce null (IndexIn) or false (IsIn) values in the output.
/// If false, any null in `value_set` is successfully matched in
/// the input.
```

Or, to put it another way.  The behavior we currently get (regardless of skip_nulls) is:

| Input Value | Lookup Set | Skip Nulls | Result |
| --- | --- | --- | --- |
| 5 | [1, NULL] | any value | false |
| 5 | [5, NULL] | any value | true |
| NULL | [5] | any value | false |
| NULL | [5, NULL] | true | false |
| NULL | [5, NULL] | false | true |

This makes perfect sense for the "null as sentinel" understanding of null.  However, SQL typically takes a "null is unknown" approach.  In which case we would want the following behavior:

| Input Value | Lookup Set | Skip Nulls | Result |
| --- | --- | --- | --- |
| 5 | [1, NULL] | any value | NULL |
| 5 | [5, NULL] | any value | true |
| NULL | [5] | any value | NULL |
| NULL | [5, NULL] | any value | NULL |

### Component(s)

C++

### PR DESCRIPTION
**Pull request by R-JunmingChen**
### Rationale for this change
As #36420 says, we want add an sql-compatible `is_in` variant, which has a different logic handling Null. After a dicussion with @ianmcook and @bkietz, we decide to support an enum option `null_matching_behavior` for SetLookup, which actually adds two semantics of null handling for `is_in` and doesn't add an new behavior for `index_in`.

The enum option `null_matching_behavior` will replace `skip_nulls` in the future.

### What changes are included in this PR?
Add an enum parameter `null_matching_behavior` for SetLookupOptions.

### Are these changes tested?
Two kinds of tests are implemented
- Replace default parameter with `null_matching_behavior` instead of `skip_nulls` for `is_in` and `index_in` tests 
- Add tests for `NullMatchingBehavior::EMIT_NULL` and `NullMatchingBehavior::INCONCLUSIVE` for `is_in`

Besides, since the `skip_nulls` is not deprecated now, I still preserve the old tests with `skip_nulls`. When the `skip_nulls` is totally deprecated, we can replace the test parameter `skip_nulls=false` with `null_matching_behavior=MATCH` and `skip_nulls=true` with `null_matching_behavior=SKIP` for these old tests.

### Are there any user-facing changes?
No. Currently we support backward compatibility. In the future, we plan to replace `skip_nulls` with `null_matching_behavior` completely.

* Closes: #36420

### CODE REVIEW

**Comment by pitrou** [comment]
Why not improve `SetLookupOptions` instead to allow selecting the desired semantics?

File: cpp/src/arrow/compute/kernels/scalar_set_lookup.cc
Comment (js8544) [inline cpp/src/arrow/compute/kernels/scalar_set_lookup.cc:418]: This should probably be `Set()`? null_bitmap[i] = True when the slot is valid, not when it's null.

**Comment by ianmcook** [comment]
> Why not improve `SetLookupOptions` instead to allow selecting the desired semantics?

+1. When possible, we typically prefer to add new options to existing functions instead of adding new functions.

**Comment by R-JunmingChen** [comment]
> Why not improve SetLookupOptions instead to allow selecting the desired semantics?

Hi, here are my two considerations:
1. The `null_handling` of all the kernels of `is_in` is `NullHandling::OUTPUT_NOT_NULL`. While a sql-compatible `is_in` can have `NULL` outputs.
2. It's a bit weird to add an option like `sql_compatible` specific for `is_in` in `SetLookupOptions`.  Since `SetLookupOptions` is used by `is_in` and `index_in`.

But these two considerations are not strict obstacles for using an option instead. For 1, we can set `null_handling` to `NullHandling::COMPUTED_PREALLOCATE` and init the null bitmap to `true` in exec for un-sql-compatible `is_in`. And the new option could be named as `sql_compatible`, which is only used by `is_in`. 

If you think the solution is ok or have a better solution, I will change code to an new option version soon.

**Comment by R-JunmingChen** [comment]
Since no further feedback, I plan to transfer to an new option version as I have commented before. Hi, @pitrou @ianmcook Do you think it's ok?

**Comment by ianmcook** [comment]
@R-JunmingChen see my comments at https://github.com/apache/arrow/issues/36420#issuecomment-1662594221.

Based on that, here is what I recommend:


In `SetLookupOptions`, replace the bool option `skip_nulls` with an enum option:

**`enum NullMatchingBehavior`**
How to match null values.
_Values:_
- **`enumerator MATCH`**
   Any null in `value_set` is successfully matched to null values in the lookup set
- **`enumerator SKIP`**
   Any null in `value_set` is ignored and nulls in the input produce null (`IndexIn`) or false (`IsIn`) values in the output.
-  **`enumerator EMIT_NULL`**
   Any null in `value_set` results in a null in the output.
- **`enumerator INCONCLUSIVE`**
   Any null in `value_set` results in a null in the output. Any null in the lookup set is treated as an unknown value that cannot conclusively match values in `value_set` and any value in `value_set` that cannot be conclusively matched results in a null in the output. This is consistent with the behavior of `IN` in SQL.


For backward compatibility:
- Default to `MATCH` (consistent with the current default `skip_nulls = false`)
- Continue to accept the existing boolean argument `skip_nulls` and respect it if the user does not use the new enum-based option. `false` translates to `MATCH` and `true` translates to `SKIP`.

I believe this should work with both `is_in` and `index_in`. 

@westonpace does this look OK to you?

**Comment by R-JunmingChen** [comment]
Hi, @ianmcook. I think it's a good solution to use an enum option , but I still have one question.

Why do we want to support `EMIT_NULL`?

**Comment by ianmcook** [comment]
Thanks @R-JunmingChen 

> 1. Why do we want to support `EMIT_NULL`? `MATCH`, `SKIP` and `INCONCLUSIVE` seem sufficient.

It is nice to have `EMIT_NULL` for completeness. There are some cases where the user might want that behavior—for example, to match the behavior of DuckDB's `list_contains` function like I describe here: https://github.com/apache/arrow/issues/36420#issuecomment-1662660149

> 2. Since we want support `skip_nulls` for backward compatibility, we need an `enumerator UNSET` for `NullMatchingBehavior`, which means we only use `skip_nulls` instead of `NullMatchingBehavior`, and make it default for the enum option. Cause we can't translate when user set the `skip_nulls` via the way like `setLookupOptions.skip_nulls = false`.

I think it is possible to define 2 explicit public functions of the class `SetLookupOptions`:
1. `SetLookupOptions(Datum value_set, bool skip_nulls = false)` (same as currently)
2. `SetLookupOptions(Datum value_set, NullMatchingBehavior null_matching_behavior = NullMatchingBehavior::MATCH)`

I think that will give backward compatibility.

**Comment by R-JunmingChen** [comment]
>I think it is possible to define 2 explicit public functions of the class SetLookupOptions:
>
>SetLookupOptions(Datum value_set, bool skip_nulls = false) (same as currently)
>SetLookupOptions(Datum value_set, NullMatchingBehavior null_matching_behavior = NullMatchingBehavior::MATCH)

Do you mean we remove the member variable `skip_nulls` and make it a parameter in constructor function and then translate the `skip_nulls` to `NullMatchingBehavior` ?

I think this way will affect some users who set `skip_nulls` directly via member variable. Does it ok?

If we don't remove member variable `skip_nulls, the translation problem still exists. For example:

```
SetLookupOptions option(value_set);
option.skip_nulls = true;
```

**Comment by ianmcook** [comment]
> > I think it is possible to define 2 explicit public functions of the class SetLookupOptions:
> > SetLookupOptions(Datum value_set, bool skip_nulls = false) (same as currently)
> > SetLookupOptions(Datum value_set, NullMatchingBehavior null_matching_behavior = NullMatchingBehavior::MATCH)
> 
> Hi @ianmcook, do you mean we remove the member variable `skip_nulls` and make it a parameter in constructor function and then translate the `skip_nulls` to `NullMatchingBehavior` ?
> 
> I think this way will affect some users who set `skip_nulls` directly via member variable. Does it ok?

Yes, I think this is OK. We can mark this as a breaking change if needed, but I think it should not affect most users. @westonpace or @bkietz can you please confirm?

**Comment by bkietz** [comment]
I would not expect there to be many direct users of the `skip_nulls` data member, but disentangling the corresponding cython will be annoying. Since it's a breaking change, we *could* compromise with:

```c++
/// Options for IsIn and IndexIn functions
class ARROW_EXPORT SetLookupOptions : public FunctionOptions {
 public:
  enum NullMatchingBehavior { MATCH, SKIP, EMIT_NULL, INCONCLUSIVE };

  explicit SetLookupOptions(Datum value_set, NullMatchingBehavior = MATCH);
  explicit SetLookupOptions(Datum value_set, bool skip_nulls);
  SetLookupOptions();
  static constexpr char const kTypeName[] = "SetLookupOptions";

  /// The set of values to look up input values into.
  Datum value_set;

  // will be overridden by skip_nulls if that is explicitly assigned
  NullMatchingBehavior null_matching_behavior;

  // DEPRECATED(use null_matching_behavior instead)
  std::optional<bool> skip_nulls;
};
```

with this, `opts.skip_nulls = false` will still work and operate as expected and (I think) cython changes will be minimal until we're ready fully to accommodate null_matching_behavior in python.

**Comment by R-JunmingChen** [comment]
I will implement this version ASAP. Let me turn this PR to draft temporarily

File: c_glib/arrow-glib/compute.cpp
Comment (kou) [inline c_glib/arrow-glib/compute.cpp:6469]:
```suggestion
  auto skip_nulls = (arrow_options->skip_nulls.has_value() && arrow_options->skip_nulls.value());
```

File: c_glib/arrow-glib/compute.cpp
Comment (kou) [inline c_glib/arrow-glib/compute.cpp:3408]:
```suggestion
  auto skip_nulls = (options.skip_nulls.has_value() && options.skip_nulls.value());
```

File: c_glib/arrow-glib/compute.cpp
Comment (kou) [inline c_glib/arrow-glib/compute.cpp:3353]:
```suggestion
    g_value_set_boolean(value, options->skip_nulls.has_value() && options->skip_nulls.value());
```

**Review by bkietz — CHANGES_REQUESTED** [review]
a few minor nits

File: cpp/src/arrow/compute/api_scalar.cc
Comment (R-JunmingChen) [reply cpp/src/arrow/compute/api_scalar.cc:572]: Indeed, we can remove it

File: cpp/src/arrow/compute/api_scalar.h
Comment (R-JunmingChen) [reply cpp/src/arrow/compute/api_scalar.h:298]: Sure, have updated

File: cpp/src/arrow/compute/api_scalar.h
Comment (mapleFU) [inline cpp/src/arrow/compute/api_scalar.h:274]: Would `ARROW_DEPRECATED("will be removed after removing of skip_nulls")` better? Or this function would be removed later?

File: cpp/src/arrow/compute/api_scalar.h
Comment (mapleFU) [inline cpp/src/arrow/compute/api_scalar.h:297]: Hmmm can we set `null_matching_behavior = GetNullMatchingBehavior()` in the constructor `SetLookupOptions(Datum value_set, bool skip_nulls)`, and remove `GetNullMatchingBehavior()`?

**Review by mapleFU — COMMENTED** [review]
Great work! I'm trying to understand the null handling logic here, just some naive questions about `Options`

File: cpp/src/arrow/compute/api_scalar.h
Comment (R-JunmingChen) [reply cpp/src/arrow/compute/api_scalar.h:297]: No, since `skip_nulls` can be set via its member variable like this.
```
SetLookupOptions option(value_set);
option.skip_nulls = true;
```
The `null_matching_behavior` set in the constructor has a possibility to contradict with `skip_nulls`. To support backward compatibility, we assign `skip_nulls` with a higher priority.

File: cpp/src/arrow/compute/api_scalar.h
Comment (R-JunmingChen) [reply cpp/src/arrow/compute/api_scalar.h:274]: Let's replace

File: cpp/src/arrow/compute/api_scalar.h
Comment (mapleFU) [reply cpp/src/arrow/compute/api_scalar.h]: Got it, thanks!

File: cpp/src/arrow/compute/api_scalar.h
Comment (R-JunmingChen) [reply cpp/src/arrow/compute/api_scalar.h:276]: I misunderstood, these functions would be removed later indeed.

Comment (mapleFU) [reply cpp/src/arrow/compute/api_scalar.h:276]: Oh I got it, sorry for misleading

File: cpp/src/arrow/compute/api_scalar.cc
Comment (R-JunmingChen) [reply cpp/src/arrow/compute/api_scalar.cc]: It's necessary that `null_matching_behavior` has a initialized value. Without initialization, `null_matching_behavior` can't be deserialized in the expression using `is_in`/`index_in`.

**Comment by R-JunmingChen** [comment]
Hi, @bkietz @ianmcook, please go on the review when you have time

File: cpp/src/arrow/compute/kernels/scalar_set_lookup.cc
Comment (bkietz) [inline cpp/src/arrow/compute/kernels/scalar_set_lookup.cc:280]: Would you mind modifying this visit overload to:
```diff
-  Status Visit(const DataType& type) {
-    DCHECK_EQ(type.id(), Type::NA);
+  Status Visit(const DataType& type) { DCHECK(false) << "IndexIn " << type; }
+  Status Visit(const NullType&) {
```

File: cpp/src/arrow/compute/kernels/scalar_set_lookup.cc
Comment (bkietz) [inline cpp/src/arrow/compute/kernels/scalar_set_lookup.cc:406]: Same here

File: cpp/src/arrow/compute/api_scalar.cc
Comment (bkietz) [inline cpp/src/arrow/compute/api_scalar.cc:371]: I think it'd be worthwhile to change this to only report/accept assignment to null_matching_behavior:
```suggestion
    CoercedDataMember("null_matching_behavior", &SetLookupOptions::null_matching_behavior,
      &SetLookupOptions::GetNullMatchingBehavior));
```

(This only affects how we deal with SetLookupOptions in reflected contexts like stringification; we'll still be able to assign to `skip_nulls`.) This can be enabled by putting the following into `reflection_internal.h`:

```c++
template <typename C, typename T>
struct CoercedDataMemberProperty {
  using Class = C;
  using Type = T;

  constexpr const Type& get(const Class& obj) const { return obj.*get_coerced_(); }

  void set(Class* obj, Type value) const { (*obj).*ptr_for_set_ = std::move(value); }

  constexpr std::string_view name() const { return name_; }

  std::string_view name_;
  Type Class::*ptr_for_set_;
  Type (Class::*get_coerced_)() const;
};

template <typename Class, typename Type>
constexpr CoercedDataMemberProperty<Class, Type> CoercedDataMember(std::string_view name,
                                                                   Type Class::*ptr,
                                                                   Type (Class::*get)()
                                                                       const) {
  return {name, ptr, get};
}
```

**Review by bkietz — CHANGES_REQUESTED** [review]
This is looking good, thanks for working on this

File: cpp/src/arrow/compute/kernels/scalar_set_lookup_test.cc
Comment (ianmcook) [inline cpp/src/arrow/compute/kernels/scalar_set_lookup_test.cc:247]: For completeness, I think we should:
- repeat each `skip_null=false` check using `null_matching_behavior=SetLookupOptions::NullMatchingBehavior::MATCH`
- repeat each `skip_null=true` check using `null_matching_behavior=SetLookupOptions::NullMatchingBehavior::SKIP`

**Comment by ianmcook** [comment]
Looks great, thanks for your effort and persistence @R-JunmingChen!

I built Arrow from your fork and created a standalone example to test this. Everything seems to be working well. https://gist.github.com/ianmcook/25bb0c82b8af4fc1bde0768863607355.

I checked whether anything needs to change on the `compute.rst` docs page. No changes are required there, I think.

File: cpp/src/arrow/compute/api_scalar.h
Comment (R-JunmingChen) [reply cpp/src/arrow/compute/api_scalar.h]: Have moved

File: cpp/src/arrow/compute/kernels/scalar_set_lookup.cc
Comment (R-JunmingChen) [reply cpp/src/arrow/compute/kernels/scalar_set_lookup.cc:284]: Sure, I have modified to that. Besides, for `Status Visit(const DataType& type)`, I add a return `Status::NotImplemented("IndexIn has no implementation with value type ", type)`

File: cpp/src/arrow/compute/kernels/scalar_set_lookup.cc
Comment (R-JunmingChen) [reply cpp/src/arrow/compute/kernels/scalar_set_lookup.cc:410]: Sure

File: cpp/src/arrow/compute/kernels/scalar_set_lookup_test.cc
Comment (R-JunmingChen) [reply cpp/src/arrow/compute/kernels/scalar_set_lookup_test.cc:251]: Sure, have added

File: cpp/src/arrow/compute/api_scalar.cc
Comment (R-JunmingChen) [reply cpp/src/arrow/compute/api_scalar.cc:371]: image
Hi, here's an error I have no idea how to handle it when I put it to `reflection_internal.h`. Do you have any solution about it?

Comment (R-JunmingChen) [reply cpp/src/arrow/compute/api_scalar.cc]: It works!

**Review by ianmcook — APPROVED** [review]
This looks good to me in terms of its API and docs. I defer to @bkietz to give a technical approval.

It looks like Python tests are failing. Can you troubleshoot that please? Thanks

**Review by bkietz — APPROVED** [review]
LGTM.

CI failures seem unrelated:

[Dev failure](https://github.com/apache/arrow/actions/runs/6251492303/job/16972734733?pr=36739#step:7:65) looks like https://github.com/apache/arrow/issues/37803

[Appveyor failure](https://ci.appveyor.com/project/ApacheSoftwareFoundation/arrow/builds/48085264#L2737) looks like https://github.com/apache/arrow/issues/37790

I've restarted the python test (which seemed to fail inside pip). If that passes I think we can merge

**Comment by ianmcook** [comment]
@ursabot please benchmark

**Comment by R-JunmingChen** [comment]
The Python failure looks like #37803.

**Comment by ianmcook** [comment]
I'm not sure why Conbench isn't working as expected here. The benchmarks should run after this merges and we can check for regressions there.

**Comment by austin3dickey** [comment]
>I'm not sure why Conbench isn't working as expected here.

Sorry, this looks to be a bug with the "ursabot please benchmark" workflow. Will fix today.

**Comment by austin3dickey** [comment]
That should be fixed. If you push an empty commit (https://github.com/voltrondata-labs/arrow-benchmarks-ci/issues/111) and then re-run the command, you can try benchmarks again. Else, yeah, they will be run post-merge.

**Comment by ianmcook** [comment]
@bkietz if you want to go ahead and merge when you're comfortable with this, we can just look at the benchmarks post-merge

**Pull request merged**
